### PR TITLE
The step exports invalid packages when using go modules [PLAT-319]

### LINKED ISSUE #2: Go 1.15: when calling "go", "list", "./..." it'll print "go: downloading ..." for packages which aren't downloaded yet
This means that this step with Go 1.15 for a repo which uses go mod and don't store the dependencies in the repo (in `vendor` dir) will produce an output like:

```
List of packages:
✓ go: downloading github.com/joho/godotenv v1.3.0
✓ go: downloading github.com/kelseyhightower/envconfig v1.4.0
✓ go: downloading github.com/pkg/errors v0.9.1
✓ github.com/bitrise-io/my-repo
✓ github.com/bitrise-io/my-repo/submoduleA
✓ github.com/bitrise-io/my-repo/submoduleB
```

instead of the desired:

```
List of packages:
✓ github.com/bitrise-io/my-repo
✓ github.com/bitrise-io/my-repo/submoduleA
✓ github.com/bitrise-io/my-repo/submoduleB
```

### PR DESCRIPTION
**Pull request by sylank**
### Files to review (but please scroll over the changed files' name)
* main_test.go
* main.go
* gotool/gotool_test.go
* gotool/gotool.go
* gotool/commander.go
* bitrise.yml
* Gopkg.toml

### Technical description
I wrote tests and refactored the whole code
I updated the dependencies
I fixed the issue with a filtering

The output logging is changed a bit because of the filtering logic.

### Description of the issue
Go list step calls go list command to see the packages in a go project, but when the project has modules in it, it will install the dependencies first, which will mean an additional line of text that the step understands as a package.

### Context (optional)
When using on a go project with modules.

### Steps to reproduce:
Run the step on a go project that has modules.

### Actual result:
Dependency installation lines understood as package.

### Expected result:
Only real packages should be parsed and exported to the output package list env.

### CODE REVIEW

File: main_test.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_filterPackages(t *testing.T) {
+	a := []string{"a", "b", "c"}
+	b := []string{"a", "b"}
```
Comment (zsolt-marta-bitrise) [inline main_test.go]: Could you please name these more specifically? like `original` and `excludes`

File: main_test.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+	diffResult := filterLines(a, b)
+
+	filteredLines := FilteredLines{}
+	filteredLines.Expected = []string{"c"}
+
+	assert.Equal(t, filteredLines.Expected, diffResult.Expected)
```
Comment (zsolt-marta-bitrise) [inline main_test.go:19]: I'd add also the assertion for the `Removed` items

File: gotool/gotool_test.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+func (c TestCommander) ExecuteCommand(command string, args ...string) (string, error) {
+	if c.ExecuteCommandFn == nil {
+		panic("You have to override TestCommander.ExecuteCommand function in tests")
+	}
+	return c.ExecuteCommandFn(command, args...)
+}
+
+func Test_ListPackages(t *testing.T) {
+	testCases := []struct {
```
Comment (zsolt-marta-bitrise) [inline gotool/gotool_test.go]: This method is a bit bloaty and hard to read. I suggest moving out the test cases definition to clear it up

File: gotool/gotool_test.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type TestCommander struct {
+	ExecuteCommandFn func(command string, args ...string) (string, error)
```
Comment (zsolt-marta-bitrise) [inline gotool/gotool_test.go]: While this looks a pretty good custom solution, I'd urge you to take a look at [testify](https://github.com/stretchr/testify), which provides a standard way of mocking objects.
For an example usage, take a look here: https://github.com/bitrise-io/beitris/pull/38/files#diff-8d6851f3ab5c8373ede6234121198af8

File: gotool/gotool.go [hunk 5]
```diff
@@ -7,6 +7,19 @@ import (
 	"github.com/bitrise-io/go-utils/command"
 )
 
+// CommandExecutor ...
+type CommandExecutor struct{}
+
+// ExecuteCommand ...
+func (c CommandExecutor) ExecuteCommand(stringCommand string, args ...string) (string, error) {
```
Comment (zsolt-marta-bitrise) [inline gotool/gotool.go:19]: Can we integrate this into `"github.com/bitrise-io/go-utils/command"` somehow? I'm afraid that these additional layers will get added to more repos and we forget about them and that leaves a lot of inconsistent duplicate code.
If not, can we check if we really need this new type?

**Review by zsolt-marta-bitrise — CHANGES_REQUESTED** [review]
Looks ok 👍 I added a couple of improvement ideas and questions

**Comment by zsolt-marta-bitrise** [comment]
Thanks for the description, especially the files to cover 😎 ! Can you please also include the jira ticket 🙏 ?

**Comment by sylank** [comment]
> Thanks for the description, especially the files to cover 😎 ! Can you please also include the jira ticket 🙏 ?

We don't copy the Jira card ID because this is an open-source project. It won

**Comment by sylank** [comment]
> Thanks for the description, especially the files to cover 😎 ! Can you please also include the jira ticket 🙏 ?

We don't copy the Jira card URL into the open-source project's PR because it doesn't give any useful information for the outside contributors and our tracking system is internal.
We are planning to create an external(open) issue board.
I set the Jira card ID and the GitHub integration will connect the card and the PR.

File: gotool/gotool.go (hunk 5 above)
Comment (zsolt-marta-bitrise) [reply gotool/gotool.go:19]: Let's not block this PR for now, but in the future we should come up with a common solution for this

**Review by zsolt-marta-bitrise — APPROVED** [review]
Thanks for the changes 🥇

File: bitrise.yml [hunk 6]
```diff
@@ -3,74 +3,74 @@ default_step_lib_source: https://github.com/bitrise-io/bitrise-steplib.git
...
-        - content: |
-            #!/usr/bin/env bash
-            set -e
-            echo '$BITRISE_GO_PACKAGES='"($BITRISE_GO_PACKAGES)"
-            EXPECTED=github.com/bitrise-steplib/steps-go-list
-            if [[ "$BITRISE_GO_PACKAGES" != "$EXPECTED" ]] ; then
-              echo "should be: $EXPECTED"
-              exit 1
-            fi
+      - path::./:
```
Comment (trapacska) [inline bitrise.yml:15]: please add another test in which it is reproducing the original issue

**Review by trapacska — CHANGES_REQUESTED** [review]

**Review by trapacska — APPROVED** [review]

**Pull request merged**